File: LeetCode_Algorithm/241(Different Ways to Add Parentheses)/241.py
from typing import List
from unittest import expectedFailure
class Solution:
    def __init__(self):
        self.memo = {}
    def diffWaysToCompute(self, expression: str) -> List[int]:
        # sol1 55ms 61%
        if expression.isdigit():
            return [int(expression)]
        # memo를 사용함으로써 똑같은 계산 하는 것을 줄임
        if expression in self.memo:
            return self.memo[expression]
        ans = []
        for i,c in enumerate(expression):
            if c in "+-*":
                left = self.diffWaysToCompute(expression[:i])
                right = self.diffWaysToCompute(expression[i+1:])
                for j in left:
                    for k in right:
                        ans.append(self.cal(j,c,k))
        self.memo[expression] = ans
        return ans
    def cal(self, m, op, n):
        if op == '+':
            return m+n
        elif op == '-':
            return m-n
        else:
            return m*n
        
print(Solution().diffWaysToCompute("2-1-4"))
# print(Solution().diffWaysToCompute("2-1-1"))
# print(Solution().diffWaysToCompute("2*3-4*5"))

File: LeetCode_Algorithm/241(Different Ways to Add Parentheses)/241.js
memo = new Map()
var diffWaysToCompute = function(input) {
  // sol1 71ms 94%
  // 어떤 수가 NaN 인지 체크할 경우 
  // == NaN 이 아닌 isNaN 함수 사용
  if (!isNaN(input)) {
    return [Number(input)]
  }
  if (memo[input]) {
    return memo[input]
  }
  let ans = [];
  for (let i = 0; i < input.length; i++) {
    if (input[i] == '+' || input[i] == '-' || input[i] == '*') {
      let left = diffWaysToCompute(input.slice(0, i));
      let right = diffWaysToCompute(input.slice(i+1));
      for (let l of left) {
        for (let r of right) {
          ans.push(cal(l,input[i],r))
          }
        }
      }
  }
  memo[input] = ans
  return ans
};
var cal = function(m,op,n) {
  if (op == '+'){
    return m+n
  } else if (op == '-') {
    return m-n
  } else {
    return m*n
  }
}
// console.log(diffWaysToCompute("2-1-1"))
console.log(diffWaysToCompute("2*3-4*5"))
